File: javascript/code-challenges/tree-intersection/binary-search-tree.js
"use strict";

const BinaryTree = require("./tree.js");
const Node = require("./node.js");

class BinarySearchTree extends BinaryTree {
  add(value) {
    let node = new Node(value);
    if (!this.root) {
      this.root = node;
      return this;
    }
    let current = this.root;
    //pointer for the previous node so we can assign left/right after traversal
    let prev = null;

    //Traverses the entire tree to get to a leaf
    while (current) {
      prev = current;
      if (value < current.value) {
        current = current.left;
      } else {
        current = current.right;
      }
    }

    // Checks prev values against passed value to see if the node belongs to the left or right
    if (value < prev.value) {
      prev.left = node;
    } else {
      prev.right = node;
    }
    return this;
  }

  addR(value, current = this.root) {
    if (!this.root) {
      this.root = new Node(value);
      return this;
    }

    if (!current) {
      return new Node(value);
    }

    if (current.value < value) {
      current.right = this.addR(value, current.right);
    } else {
      current.left = this.addR(value, current.left);
    }

    return current;
  }

  contains(value) {
    if (this.root === null) {
      return false;
    }

    let current = this.root;

    while (current) {
      if (value > current.value) {
        current = current.right;
      } else if (value < current.value) {
        current = current.left;
      } else {
        return true;
      }
    }
    return false;
  }
}

module.exports = BinarySearchTree;

File: java/codechallenges/lib/src/main/java/codechallenges/trees/BinarySearchTree.java
package codechallenges.trees;

import codechallenges.linkedList.Node;

public class BinarySearchTree<T extends Comparable<? super T>> extends BinaryTree<T> {


  public Node<T> add(T value, Node<T> currentNode){
    Node<T> newNode = new Node(value);
    if(this.root == null){
      this.root = newNode;
      return newNode;
    }
    if(currentNode == null){
      return newNode;
    }
    if(newNode.value.compareTo(currentNode.value) > 0){
      currentNode.right = this.add(value, currentNode.right);
    } else if(newNode.value.compareTo(currentNode.value) < 0){
      currentNode.left = this.add(value, currentNode.left);
    }
    return currentNode;
  }

  public boolean contains(T value, Node<T> currentNode){
    if(value.compareTo(currentNode.value) == 0){return true;}
    if(value.compareTo(currentNode.value) > 0){return this.contains(value, currentNode.right);}
    if(value.compareTo(currentNode.value) < 0){return this.contains(value, currentNode.left);}
    return false;
  }
}
